autofocus/predict/process__predict_example.R

```diff
@@ -0,0 +1,217 @@
+#   2. It also assumes that the api address is at 127.0.0.1 
+#        (which should be the case)
+#   3. Assumes that your current working directory is 
+#       './GitHub/autofocus/autofocus/predict'
+#   4. Assumes that the images you are going to send to autofocus have
+#        not been preprocessed at all. 
+
+#Library requirements:
+# RCurl, jsonlite, dplyr, magick, zip, progress
+
+library(RCurl)
+library(jsonlite)
+library(magick)
+library(zip)
+library(progress)
+library(dplyr)
+
+find_image_files <- function(search_dir, 
+                             image_list = c("jpeg","jpg","bmp","png", "JPG")){
+  # Utility function to find all recursively find all image files 
+  #   starting from a directory
+  
+  # Args:
+  #   search_dir(character): the starting directory path from which to search
+  #     image_list(list): a list of acceptable file formats
+  
+  # Returns:
+  #   image_files(list): list containing the paths of all image files found. 
+  #     Each element in this list is a vector of at least 10 images. This split
+  #     is done so that the images can be zipped and sent to autofocus.
+  
+  file_list <- list.files(search_dir, recursive = TRUE, full.names = TRUE)
+  image_files <- file_list[grep(paste(image_list, collapse = "|"), file_list)]
+  image_files <- normalizePath(image_files, winslash = "/")
+  # normalize the path, then split into groups of max 10 image
+  n_groups <- ceiling(length(image_files) / 10)
+  image_files <- split(image_files,
+                       sort(rep_len(1:n_groups, length(image_files))))
+  
+  return(image_files)
+}
+
+
+process_images <- function(image_files = NULL){
+  # Utility function to preprocess images to be sent to autofocus
+  
+  # Args:
+  #   image_files(list): the output object from find_images()
+  
+  # Returns:
+  #   a list: This list has two elements:
+  #     1. zip(character): A vector of the temporary zip files to be sent to 
+  #          autofocus.
+  #     2. dict(named character): a key-value pair that links the temporary
+  #          image file to the actual file. The elements in this vector are
+  #          the names of the temporary files while the names are the full 
+  #          paths to the file names.
+  
+  if(!is(image_files, 'list'))
+    stop('image_files must be a list.')
+  
+  if(any(sapply(image_files, length)>10))
+    stop('One of the elements is image_files has > 10 images.')
+  
+  dict_list <- vector('list', length = length(image_files))
+  zip_vector <- rep(NA, length(image_files))
+  
+  cat(paste('Processing', length(unlist(image_files)), 'images...\n'))
+  
+  pb <- progress_bar$new(
+    format = "Images processed [:bar] :elapsed | eta: :eta",
+    total = length(unlist(image_files)),
+    width = 60
+  )
+
+  for(photo_group in seq.int(length(image_files))){
+    
+    file_pattern <- paste0("file_",stringr::str_pad(1:length(image_files[[photo_group]]),
+                               width = 2, pad = "0"),"_")
```

Comment:
since you reference `image_files[[photo_group]]` in a bunch of places, I think it would make the code cleaner to pull it out one time at the beginning of this for loop. Like this:

```r
for (photo_group in seq.int(length(image_files))){
    image_file_names <- image_files[[photo_group]] 
    num_files <- length(image_file_names)
   ....
}
```

Then you can reference `image_file_names` and `num_images` instead of repeatedly doing `length(image_files[[photo_group]])`.

Reply:
Sounds good. I'm going to add to this as well though. We should probably make sure that the image is not corrupt before processing it or else the number of files and the like will be off (and file resizing won't work). The corrupt images are  `0 kb` so we can just check for them between collecting the file names and figuring out the number of files. That way, it looks something like:

```R
    # get paths to files
    image_file_names <- image_files[[photo_group]]
    
    # files with 0 kb are corrupt, remove them
    file_sizes <- file.size(image_file_names)
    if(any(file_sizes == 0)){
      image_file_names <- image_file_names[-which(image_file_names == 0)]
    }
    
    # count number of files
    num_files <- length(image_file_names)
```